### PR TITLE
[REVIEW] Serializer for mdspans

### LINKED ISSUE #1017: [FEA] Tracker for mdspan serialization and copy
Following up with #752.

Design choices
* Provide efficient serializers for mdspan's with C and Fortran layouts. Mdspan's can have an arbitrary layouts which makes it hard to serialize efficiently.
* Provide a copy operator so that users can transform their mdspans into C or Fortran layouts.
  - In general, provide a generic fallback implementation (possibly suboptimal), and provide specialized versions of `copy` that are optimized for certain use cases.
* Adopt [the NumPy format](https://numpy.org/doc/1.23/reference/generated/numpy.lib.format.html) for easy inter-op with NumPy and other array libraries.
* Also expose `__array_interface__` for zero-copy serialization

TODO's

- [ ] Implement a copy operation between two mdspan's
  * [x] Implement fallback copy operation to copy elements between two mdspan's on the host memory
  * [ ] Implement fallback copy operation to copy elements between two mdspan's on the device memory
  * [ ] Implement wrappers for copying Fortran-, C-layout, or strided arrays between the host and device memory
- [ ] Implement serializer for mdspan's with Fortran and C layouts
  * [ ] Serialize to `std::ostream`
  * [ ] Serialize to container iterator
  * [ ] Serialize via `__array_interface__`

### PR DESCRIPTION
**Pull request by hcho3**
Closes #1017

Implement a serializer for mdspan using the NumPy format. Row- and column-major layouts are supported.

TODOs
- [x] Revise IVF to use the new mdspan serializer
- [x] Add gtest to cover serializing device mdspans
- [x] Implement serializer for scalars
- [x] Rename header to `raft/core/serialize.hpp`
- [x] Use `device_resources` instead of `handle_t`
- [x] Use 64-byte alignment
- [x] Serialize scalars as 0D NumPy array
- [x] Add version field to ANN serialization
- [x] Test mdspan serializer in the Python layer

### CODE REVIEW

File: cpp/include/raft/core/mdspan_serializer.hpp
Comment (cjnolet) [inline cpp/include/raft/core/mdspan_serializer.hpp:2]: What do you think about `raft/core/serialize.hpp`?

File: cpp/include/raft/spatial/knn/detail/ivf_flat_build.cuh
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/detail/ivf_flat_build.cuh:406]: It would be nice if all of these `write_*` could be moved to `raft/core/serialize.hpp` (and renamed to `serialize_*` and `deserialize_*`).

File: cpp/include/raft/core/mdspan_serializer.hpp
Comment (hcho3) [reply cpp/include/raft/core/mdspan_serializer.hpp:2]: @cjnolet Do we have serializers for other RAFT types?

Comment (cjnolet) [reply cpp/include/raft/core/mdspan_serializer.hpp:2]: Not yet, but when we do (such as serializing scalars, key/value pairs, etc...) then we can stick them in this header.

**Comment by hcho3** [comment]
@cjnolet I introduced a few macros to replace global constant variables, as the latter caused a link error due to symbol duplication. Is this okay? Macros are not hygienic and will not be scoped by namespaces. The alternative is to add new source file(s). What's the policy like for adding source files? Is RAFT meant to be mostly header only?

**Comment by cjnolet** [comment]
> Macros are not hygienic and will not be scoped by namespaces

I think what you've done should be okay since they are at least prefixed w/ `RAFT_NUMPY` thus lowering the potential for collision. 

> What's the policy like for adding source files? Is RAFT meant to be mostly header only?

RAFT's APIs are all header-only and the source files are used for 1) template specializations that users can link against to avoid recompiling expensive headers, and 2) runtime APIs that can be accessed on host. This means the serializer code needs to be header-only at its core but if we find that the functions are expensive to compile, we can always add an explicit template specialization for them.

**Comment by hcho3** [comment]
Got it, thanks for clarification

File: cpp/include/raft/core/serialize.hpp
Comment (tfeher) [inline cpp/include/raft/core/serialize.hpp:39]: Oh, now I see that here are the asserts I was missing from numpy_serializer. I think it would make sense to move them closer to the actual code which relies on these assumptions.

File: cpp/include/raft/core/serialize.hpp
Comment (tfeher) [inline cpp/include/raft/core/serialize.hpp:65]: Make sure the data has arrived
```
      handle.sync_stream();
```

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi @hcho3, thanks for this work! I am really excited to have this in raft! It looks great overall, I have just a few comments.

File: cpp/include/raft/core/detail/mdspan_numpy_serializer.hpp
Comment (hcho3) [reply cpp/include/raft/core/detail/mdspan_numpy_serializer.hpp:324]: My bad. I was looking at [an old version of the NumPy format](https://numpy.org/doc/1.13/neps/npy-format.html) and didn't realize they later [increased the alignment requirement to 64](https://github.com/numpy/numpy/pull/9025). Let me update the code.

**Review by tfeher — APPROVED** [review]
Thanks @hcho3 for the update! The PR looks good to me.

**Comment by hcho3** [comment]
Can I get another round of review?

File: cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh
Comment (achirkin) [inline cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh:1376]: We've already had a problem with ivf-pq once, that `index` members changed but we forgot to update the load/save functions. Would be nice to detect these changes when updating the code, but I'm not aware of any good way to do this in C++.
As a very crude approximation, what do you think about adding something like this to `load` and `save` functions?
```C++
static_assert(sizeof(ivf_pq::index<uint64_t>) == {literal}, "The size of the index struct has changed since the last update; paste in the new size and consider updating the save/load logic");
```

File: cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh
Comment (achirkin) [inline cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh:1375]: Maybe also add a `constexpr` while you're at it?

File: cpp/include/raft/core/mdspan.hpp
Comment (achirkin) [inline cpp/include/raft/core/mdspan.hpp:37]: This seems to be only used in tests, is it? If so, do we need to re-export this definition?

**Review by achirkin — CHANGES_REQUESTED** [review]
I'm excited finally to get this in raft! A few minor things in the in the comments.

Also, I'm concerned a little bit regarding the naming. Atm we use `serialize_xx`, `deserialize_xx` and also `load`,`save`. This looks a bit inconsistent. Would you consider changing all of these to `serialize_xx`/`deserialize_xx`? Or even just `serialize`/`deserialize` (and rely on overload machinery)?

File: cpp/include/raft/core/detail/mdspan_numpy_serializer.hpp
Comment (hcho3) [reply cpp/include/raft/core/detail/mdspan_numpy_serializer.hpp:119]: Good point, we'd need to create a specialization specifically for `half` type.

**Comment by hcho3** [comment]
> Would you consider changing all of these to `serialize_xx`/`deserialize_xx`? Or even just `serialize`/`deserialize` (and rely on overload machinery)?

I'm in favor of `serialize_xx`/`deserialize_xx`, because using the same `serialize` name for both scalars and mdspans may confuse the compiler and cause a surprising behavior in overload resolution. That is, when we declare
```
template <typename T>
void serialize(std::ostream& os, const T& value)
```
to serialize scalars, the function may be accidentally called with an mdspan (by setting `T=mdspan<...>`). 

I have rename all functions to `serialize_xxx` / `deserialize_xxx`.

File: cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh
Comment (hcho3) [reply cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh:1403]: Done

**Review by achirkin — APPROVED** [review]
LGTM!

**Review by cjnolet — APPROVED** [review]
I’m pre-approving so we don’t need to hold up the release for these but just a couple small things we should probably open a follow-on for (in 23.04)

File: cpp/include/raft/core/serialize.hpp
Comment (divyegala) [inline cpp/include/raft/core/serialize.hpp:35]: Can we use the base-class `raft::resources` for all these functions? `device_resources` is not needed with `host_mdspan` overloads

**Review by divyegala — APPROVED** [review]
Pre-approving, the PR looks really great! Just one comment, could you please file an issue or add a TODO for my comment to be tackled later?

**Review by raydouglass — APPROVED** [review]

**Pull request merged**
